File: merry360x_mobile/android-app/app/src/main/java/com/merry360x/mobile/viewmodel/HomeViewModel.kt
package com.merry360x.mobile.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.merry360x.mobile.data.CityWithCount
import com.merry360x.mobile.data.Listing
import com.merry360x.mobile.data.SupabaseApi
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class CitySection(
    val city: String,
    val count: Int,
    val listings: List<Listing>
)

data class HomeUiState(
    val loading: Boolean = false,
    val listings: List<Listing> = emptyList(),
    val tours: List<Listing> = emptyList(),
    val cars: List<Listing> = emptyList(),
    val events: List<Listing> = emptyList(),
    val cities: List<CityWithCount> = emptyList(),
    val citySections: List<CitySection> = emptyList(),
    val error: String? = null,
)

class HomeViewModel(
    private val api: SupabaseApi,
) : ViewModel() {
    private val _state = MutableStateFlow(HomeUiState())
    val state: StateFlow<HomeUiState> = _state.asStateFlow()

    private fun extractRegion(location: String): String {
        val tokens = location
            .split(",")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

        if (tokens.isEmpty()) return "Other"
        if (tokens.size == 1) return tokens.first()

        val broadValues = setOf(
            "kigali", "rwanda", "rw", "gasabo", "nyarugenge", "kicukiro",
            "rubavu", "musanze", "huye", "bugesera", "nyanza", "karongi"
        )

        return if (broadValues.contains(tokens.first().lowercase())) {
            tokens[1]
        } else {
            tokens.first()
        }
    }

    fun load() {
        viewModelScope.launch {
            _state.value = _state.value.copy(loading = true, error = null)
            try {
                val stayDeferred = async { api.fetchFeaturedListings(limit = 1000) }
                val toursDeferred = async { api.fetchTours(limit = 300) }
                val carsDeferred = async { api.fetchCars(limit = 300) }
                val eventsDeferred = async { api.fetchEvents(limit = 300) }

                val listings = stayDeferred.await()
                val tours = toursDeferred.await()
                val cars = carsDeferred.await()
                val events = eventsDeferred.await()

                val grouped = listings.groupBy { listing -> extractRegion(listing.location) }

                val sections = grouped
                    .map { (region, regionListings) ->
                        CitySection(region, regionListings.size, regionListings)
                    }
                    .sortedWith(compareByDescending<CitySection> { it.count }.thenBy { it.city.lowercase() })

                val cities = sections.map { CityWithCount(it.city, it.count) }
                
                _state.value = _state.value.copy(
                    loading = false,
                    listings = listings,
                    tours = tours,
                    cars = cars,
                    events = events,
                    cities = cities,
                    citySections = sections
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(loading = false, error = e.message ?: "Failed to load")
            }
        }
    }
}

File: merry360x_mobile/ios-app/Merry360xMobile/ViewModels/HomeViewModel.swift
import Foundation

struct CitySection: Identifiable {
    let id = UUID()
    let city: String
    let count: Int
    var listings: [Listing] = []
}

@MainActor
final class HomeViewModel: ObservableObject {
    @Published var listings: [Listing] = []
    @Published var tours: [Listing] = []
    @Published var cars: [Listing] = []
    @Published var events: [Listing] = []
    @Published var citySections: [CitySection] = []
    @Published var cities: [(city: String, count: Int)] = []
    @Published var loading = false
    @Published var errorMessage: String?

    private let service = SupabaseService()

    private func extractRegion(from location: String) -> String {
        let tokens = location
            .split(separator: ",")
            .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        if tokens.isEmpty { return "Other" }
        if tokens.count == 1 { return tokens[0] }

        let broadValues = Set([
            "kigali", "rwanda", "rw", "gasabo", "nyarugenge", "kicukiro",
            "rubavu", "musanze", "huye", "bugesera", "nyanza", "karongi"
        ])

        let first = tokens[0].lowercased()
        if broadValues.contains(first) {
            return tokens[1]
        }

        return tokens[0]
    }

    func load() async {
        guard let service else {
            errorMessage = "MobileConfig is not set. Add Supabase URL + anon key first."
            return
        }

        loading = true
        errorMessage = nil
        do {
            async let stayTask = service.fetchFeaturedListings(limit: 1000)
            async let toursTask = service.fetchTours(limit: 300)
            async let carsTask = service.fetchCars(limit: 300)
            async let eventsTask = service.fetchEvents(limit: 300)

            let stayRows = try await stayTask
            let tourRows = try await toursTask
            let carRows = try await carsTask
            let eventRows = try await eventsTask

            listings = stayRows
            tours = tourRows
            cars = carRows
            events = eventRows

            var grouped: [String: [Listing]] = [:]
            for listing in stayRows {
                let region = extractRegion(from: listing.location)
                grouped[region, default: []].append(listing)
            }

            let sortedSections = grouped
                .map { key, value in
                    CitySection(city: key, count: value.count, listings: value)
                }
                .sorted { lhs, rhs in
                    if lhs.count == rhs.count {
                        return lhs.city.localizedCaseInsensitiveCompare(rhs.city) == .orderedAscending
                    }
                    return lhs.count > rhs.count
                }

            citySections = sortedSections
            cities = sortedSections.map { (city: $0.city, count: $0.count) }
        } catch {
            errorMessage = error.localizedDescription
        }
        loading = false
    }
}
